man/man1/setclm.1
.TH setclm 1  "USER COMMANDS"
.SH NAME
setclm \- set value for specified record in header of clm files for LPJmL
.SH SYNOPSIS
.B setclm
{id|version|order|firstyear|nyear|firstcell|ncell|nbands|nstep|scalar|type|cellsize|cellsize_lon|cellsize_lat|timestep} \fIvalue\fP \fIclmfile\fP

.SH DESCRIPTION
Program sets file header and data stored in clm data files for LPJmL.
.SH OPTIONS
.TP
{id|version|order|firstyear|nyear|firstcell|ncell|nbands|nstep|scalar|type|cellsize_lon|cellsize_lat|timestep}
Define record to be set.
.TP
.I value
Value written in clm file.
.TP
.I clmfile
Filename of clm data file.
.SH EXAMPLE
.TP
Set version of clm file to 2:
.B setclm
version 2 tmp.clm
.PP
.SH EXIT STATUS
.B setclm
returns a zero exit status if value in clm file could be set.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
printclm(1), printheader(1), headersize(1), cru2clm(1), grid2clm(1), cft2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ```.gitlab/merge_request_templates/default.md``` that includes several checklists

## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.
- New function `getsprintf()` added to allocate and print formatted output into string. Function replaces call to `snprintf()` function.
- Default settings for `GIT_HASH` and `GIT_REPO` added.
- Duplicate filenames removed from list of input/output filenames in utility `lpjfiles`.

### Added

- Option `-int` added to utility `cdf2clm`.
- Option `-latlon` added to utility `cdf2coord` in order to change the order of the CLM grid file.
- Utility `cdf2reservoir` added to convert NetCDF reservoir file into CLM file.
- Utility `reservoir2cdf` added to convert CLM reservoir file into a NetCDF file using the soil code NetCDF file.
- Macro `NETCDF_INPUT` added in `lpjml_config.cjson` to enable NetCDF input.
- Reservoir, irrigation neighbor and drainage data can now be in NetCDF format:
```java
"drainage" :           { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/drainage.nc"},
"neighb_irrig" :       { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/neighb_irrig.nc"},
"river" :              { "fmt" : "cdf", "var" : "riverlen", "name" : "cru_netcdf/drainage.nc"},
"reservoir" :          { "fmt" : "cdf", "var" : "year", "name" : "cru_netcdf/reservoir.nc"},
"capacity_reservoir" : { "fmt" : "cdf", "var" : "capacity", "name" : "cru_netcdf/reservoir.nc"},
"area_reservoir" :     { "fmt" : "cdf", "var" : "area", "name" : "cru_netcdf/reservoir.nc"},
"inst_cap_reservoir" : { "fmt" : "cdf", "var" : "inst_cap", "name" : "cru_netcdf/reservoir.nc"},
"height_reservoir" :   { "fmt" : "cdf", "var" : "height", "name" : "cru_netcdf/reservoir.nc"},
"purpose_reservoir" :  { "fmt" : "cdf", "var" : "purpose", "name" : "cru_netcdf/reservoir.nc"},
```

### Fixed

- River length is now correctly read in `initdrain.c` for data in NetCDF format.
- Man page of `cvrtclm` corrected.
- Missing check for open NetCDF file added to `openclimate.c`.


## [5.9.21] - 2025-03-13

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jannes Breier (breier@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- `"coupled_host"` and `"coupled_port"` added to LPJmL configuration to specify where the coupled model is running and which port is used for communication.


## [5.9.20] - 2025-03-13

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Check for correct number of time steps added in `openclimate_netcdf.c`.
- Option `timestep` added to `setclm`.

### Changed


src/utils/setclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                       s  e  t  c  l  m  .  c                                   \n**/
/**                                                                                \n**/
/**     Program sets header of clm file                                            \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s id|version|order|firstyear|nyear|firstcell|ncell|nbands|nstep|type|scalar|cellsize|cellsize_lon|cellsize_lat|timestep value filename\n"

static Bool writeheader(FILE *file,int *header,int size,Bool swap)
{
  int i;
  if(swap)
    for(i=0;i<size;i++)
      header[i]=swapint(header[i]);
  return (fwrite(header,sizeof(int),size,file)!=size);
} /* of 'writeheader' */

int main(int argc,char **argv)
{
  FILE *file;
  Header header;
  String s;
  int version;
  char *id;
  Bool swap;
  const char *progname;
  char *endptr;
  int index,size;
  Bool setversion=FALSE;
  progname=strippath(argv[0]);
  if(argc<4)
  {
    fprintf(stderr,"Error: argument(s) missing.\n"
            USAGE,progname);
    return EXIT_FAILURE;
  }
  file=fopen(argv[3],"r+b");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s\n",argv[3],strerror(errno));
    return EXIT_FAILURE;
  }
  version=READ_VERSION;
  if(freadanyheader(file,&header,&swap,s,&version,TRUE))
  {
    fprintf(stderr,"Error reading header in '%s', perhaps not a clm file.\n",
            argv[3]);
    return EXIT_FAILURE;
  }
  id=NULL;
  if(!strcmp(argv[1],"id"))
  {
    id=argv[2];
    if(strlen(id)!=strlen(s))
    {
      fprintf(stderr,"Length of new id '%s' is different from '%s'.\n",
              id,s);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"version"))
  {
    version=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for version.\n",argv[2]);
      return EXIT_FAILURE;
    }
    setversion=TRUE;
  }
  else if(!strcmp(argv[1],"order"))
  {
    header.order=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      header.order=findstr(argv[2],ordernames,4);
      if(header.order==NOT_FOUND)
      {
        fprintf(stderr,"Invalid number '%s' for order.\n",argv[2]);
        return EXIT_FAILURE;
      }
      header.order++;
    }
  }
  else if(!strcmp(argv[1],"firstyear"))
  {
    header.firstyear=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for firstyear.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"nyear"))
  {
    header.nyear=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for nyear.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"firstcell"))
  {
    header.firstcell=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for firstcell.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"ncell"))
  {
    header.ncell=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for ncell.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"nbands"))
  {
    header.nbands=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for nbands.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"nstep"))
  {
    if(version<4)
    {
      fprintf(stderr,"Version=%d for nstep of '%s' must be >3.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.nstep=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for nstep.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"timestep"))
  {
    if(version<4)
    {
      fprintf(stderr,"Version=%d for timestep of '%s' must be >3.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.timestep=strtol(argv[2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for timestep.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"type"))
  {
    if(version<3)
    {
      fprintf(stderr,"Version=%d for type of '%s' must be >2.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    index=findstr(argv[2],typenames,5);
    if(index==NOT_FOUND)
    {
      fprintf(stderr,"Invalid argument '%s' for type.\n",argv[2]);
      return EXIT_FAILURE;
    }
    header.datatype=(Type)index;
  }
  else if(!strcmp(argv[1],"scalar"))
  {
    if(version<2)
    {
      fprintf(stderr,"Version=%d for scalar of '%s' must be >1.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.scalar=(float)strtod(argv[2],&endptr);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for scalar.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"cellsize"))
  {
    if(version<2)
    {
      fprintf(stderr,"Version=%d for cellsize of '%s' must be >1.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.cellsize_lon=header.cellsize_lat=(float)strtod(argv[2],&endptr);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for cellsize.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"cellsize_lon"))
  {
    if(version<3)
    {
      fprintf(stderr,"Version=%d for cellsize_lon of '%s' must be >2.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.cellsize_lon=(float)strtod(argv[2],&endptr);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for cellsize_lon.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else if(!strcmp(argv[1],"cellsize_lat"))
  {
    if(version<3)
    {
      fprintf(stderr,"Version=%d for cellsize_lat of '%s' must be >2.\n",version,argv[3]);
      return EXIT_FAILURE;
    }
    header.cellsize_lat=(float)strtod(argv[2],&endptr);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for cellsize_lat.\n",argv[2]);
      return EXIT_FAILURE;
    }
  }
  else
  {
    fclose(file);
    fprintf(stderr,"Invalid header item '%s'.\n"
            USAGE,argv[1],progname);
    return EXIT_FAILURE;
  }
  if(id==NULL)
    id=s;
  rewind(file);
  if(fwrite(id,strlen(id),1,file)!=1)
    return EXIT_FAILURE;
  if(swap)
    version=swapint(version);
  if(fwrite(&version,sizeof(version),1,file)!=1)
    return EXIT_FAILURE;
  if(setversion)
  {
    fclose(file);
    return EXIT_SUCCESS;
  }
  switch(version)
  {
     case 1:
       size=sizeof(Header_old)/sizeof(int);
       break;
     case 2:
       size=sizeof(Header2)/sizeof(int);
       break;
     case 3:
       size=sizeof(Header3)/sizeof(int);
       break;
     case 4:
       size=sizeof(Header)/sizeof(int);
       break;
     default:
       fprintf(stderr,"Invalid version %d, must be <5.\n",version);
       return EXIT_FAILURE;
  }
  if(writeheader(file,(int *)&header,size,swap))
    return EXIT_FAILURE;
  fclose(file);
  return EXIT_SUCCESS;
} /* of 'main' */
